README.md
# pil_printer
## Introduction
With the *pil_printer* module it is possible to print a PIL image or several PIL images to a printer.
The most common use case is to print JPG or PNG file(s) at an exact point with an exact scaling.
Most graphics programs don't support exact placement and the user has nearly always to
manually control this placement/scaling for each image.

## Restrictions
*pil_printer* only works under Windows.

## Installation
The *pil_printer* module consists of one file: pil_printer.py which should be copied to the
current working directory or any directory in the sys.path.

It might be necessary to install win32print, win32ui and Pillow.

## Usage / examples
In order to use *pil_printer*, just put
```
from pil_printer import pil_printer, spec
```
at the top of your program.

If want to print the file `a.jpg` to the printer, scaled to 10 cm wide, use
```
image = "a.jpg"
pil_printer(image, width=10)
```
The image will be printed centered on the page.

If you want to offset the image by 5 cm to the right and 10 cm up, use
```
pil_printer("a.jpg", width=10, horizontal_offset=5, vertical_offset=-10)
```
Instead of using cm as the unit, you also use inch:
```
pil_printer("a.jpg", width=4, horizontal_offset=2, vertical_offset=-4, unit="inch")
```
By default, pil_printer uses the default printer, but with the printer argument,
any printer can be selected:
```
pil_printer("a.jpg", width=10, printer="EPSON WF-3725")
```
The image parameter doesn't have to be a filename. It can also be a PIL image, either
created by loading a file or using PIL primitives:
```
from pil_printer import pil_printer, spec
from PIL import ImageDraw
im = Image.new(mode="RGBA", size=(1000, 1000), color=(255, 255, 255))
d = ImageDraw.Draw(im=im)
d.rectangle(xy=(100, 100, 900, 900), outline=(0, 0, 0))
pil_printer(im)
```
So far, we've just printed one image on a page.
Instead of the image parameter, however multiple files/images may be specified:
```
pil_printer("a.jpg", "b.jpg", width=10)
```
*This is usually not what is wanted as all images are printed with the same scaling/positioning.*

In order to print multiple images with different scaling/postioning, the spec function can be used:
```
pil_printer(spec("a.jpg", width=10, horizontal_offset=-5), spec("b.jpg", width=5, horizontal_offset=5))
```
The spec parameters override the pil_printer parameters.

**Note that it is required to specify either width or height, but never both!**

## Reference
```
def pil_printer(*image, width=None, height=None, horizontal_offset=0, vertical_offset=0, unit="cm", printer_name=None):

    Prints a PIL image to a printer
    
    Parameters
    ----------
    image : PIL image
        Image to be printed
        Or it is possible to specify many images.
        However this nearly always requires the images to be given
        as a spec, in order to set offset(s) and possibly height/width
        
    width : float 
        Width of the printed image in the given unit
        No default
        
    height : float 
        Height of  the printed image in the given unit
        No default
                    
    horizontal_offset : float
        Normally the image is printed centered on the page.
        With this parameter, the image can be offset horizontally
        Default: 0
        
    vertical_offset : float
        Normally the image is printed centered on the page.
        With this parameter, the image can be offset vertically
        Default: 0        
        
    unit : str
        Either 'cm' (default) or 'inch'

    printer : str
        Printer name to used
        Default: default printer

    Note
    ----
    One and only one of width/height has to be specified
    Only compatible with Windows platforms.
```

```
def spec(image, width=None, height=None, horizontal_offset=None, vertical_offset=None, unit=None):

    Prepares an image to be printed with custom properties
    
    Parameters
    ----------
    image : PIL image
        Image(s) to be used
        
    width : float 
        Width of the printed image in the given unit
        Defaults to width given in call to pil_printer
        
    height : float 
        Height of  the printed image in the given unit
        Defaults to height given in call to pil_printer        
                    
    horizontal_offset : float
        Defaults to horizontal_offset given in call to pil_printer
        
    vertical_offset : float
        Defaults to vertical_offset given in call to pil_printer
        
    unit : str 
        Defaults to unit given in call to pil_printer
        Must be cm or inch
        
    Returns
    -------
    spec : dict
        to be used as image in call to pil_printer
```




changelog.txt
Changelog pil_printer
=====================

Version 1.0.0  2022-07-24
-------------------------
Initial release


install pil_printer.py
from __future__ import print_function

import sys
import site
import shutil
import hashlib
import base64
from pathlib import Path
import configparser
import six
from six.moves import urllib
#import urllib.request
#import urllib.error


def _install(files, url=None):
    """
    install one file package from GitHub or current directory

    Parameters
    ----------
    files : list
        files to be installed
        the first item (files[0]) will be used as the name of the package''
        optional files should be preceded with an exclamation mark (!)

    url : str
        url of the location of the GitHub repository
        this will start usually with https://raw.githubusercontent.com/ and end with /master/
        if omitted, the files will be copied from the current directory (not GitHub)


    Returns
    -------
    info : Info instance
        info.package : name of the package installed
        info.path : name where the package is installed in the site-packages
        info.version : version of the package (obtained from <package>.py)
        info.files_copied : list of copied files

    Notes
    -----
    The program automatically makes the required __init__.py file (unless given in files) and
    <package><version>.dist-info folder with the usual files METADATA, INSTALLER and RECORDS.
    As the setup.py is not run, the METADATA is very limited, i.e. is contains just name and version.

    If a __init__.py is in files that file will be used.
    Otherwise, an __init__/py file will be generated. In thet case, if a __version__ = statement
    is found in the source file, the __version__ will be included in that __init__.py file.

    Version history
    ---------------
    version 1.0.5  2020-06-24
        Bug with removing the dist-info of packages starting with the same name fixed.

    version 1.0.4  2020-03-29
        Linux and ios versions now search in sys.path for site-packages,
        whereas other platforms now use site.getsitepackages().
        This is to avoid installation in a roaming directory on Windows.

    version 1.0.2  2020-03-07
        modified several open calls to be compatible with Python < 3.6
        multipe installation for Pythonista removed. Now installs only in site-packages

    version 1.0.1  2020-03-06
        now uses urllib instead of requests to avoid non standard libraries
        installation for Pythonista improved

    version 1.0.0  2020-03-04
        initial version

    (c)2020 Ruud van der Ham - www.salabim.org
    """

    class Info:
        version = "?"
        package = "?"
        path = "?"
        files_copied = []

    info = Info()
    Pythonista = sys.platform == "ios"
    if not files:
        raise ValueError("no files specified")
    if files[0][0] == "!":
        raise ValueError("first item in files (sourcefile) may not be optional")
    package = Path(files[0]).stem
    sourcefile = files[0]

    file_contents = {}
    for file in files:
        optional = file[0] == "!"
        if optional:
            file = file[1:]

        if url:
            try:
                with urllib.request.urlopen(url + file) as response:
                    page = response.read()

                file_contents[file] = page
                exists = True
            except urllib.error.URLError:
                exists = False

        else:
            exists = Path(file).is_file()
            if exists:
                with open(file, "rb") as f:
                    file_contents[file] = f.read()

        if (not exists) and (not optional):
            raise FileNotFoundError(file + " not found. Nothing installed.")

    version = "unknown"
    for line in file_contents[sourcefile].decode("utf-8").split("\n"):
        line_split = line.split("__version__ =")
        if len(line_split) > 1:
            raw_version = line_split[-1].strip(" '\"")
            version = ""
            for c in raw_version:
                if c in "0123456789-.":
                    version += c
                else:
                    break
            break

    info.files_copied = list(file_contents.keys())
    info.package = package
    info.version = version

    file = "__init__.py"
    if file not in file_contents:
        file_contents[file] = ("from ." + package + " import *\n").encode()
        if version != "unknown":
            file_contents[file] += ("from ." + package + " import __version__\n").encode()
    if sys.platform.startswith("linux") or (sys.platform == "ios"):
        search_in = sys.path
    else:
        search_in = site.getsitepackages()

    for f in search_in:
        sitepackages_path = Path(f)
        if sitepackages_path.name == "site-packages" and sitepackages_path.is_dir():
            break
    else:
        raise ModuleNotFoundError("can't find the site-packages folder")

    path = sitepackages_path / package
    info.path = str(path)

    if path.is_file():
        path.unlink()

    if not path.is_dir():
        path.mkdir()

    for file, contents in file_contents.items():
        with (path / file).open("wb") as f:
            f.write(contents)

    if Pythonista:
        pypi_packages = sitepackages_path / ".pypi_packages"
        config = configparser.ConfigParser()
        config.read(pypi_packages)
        config[package] = {}
        config[package]["url"] = "github"
        config[package]["version"] = version
        config[package]["summary"] = ""
        config[package]["files"] = path.as_posix()
        config[package]["dependency"] = ""
        with pypi_packages.open("w") as f:
            config.write(f)
    else:
        for entry in sitepackages_path.glob("*"):
            if entry.is_dir():
                if entry.stem.startswith(package + "-") and entry.suffix == ".dist-info":
                    shutil.rmtree(str(entry))
        path_distinfo = Path(str(path) + "-" + version + ".dist-info")
        if not path_distinfo.is_dir():
            path_distinfo.mkdir()
        with open(str(path_distinfo / "METADATA"), "w") as f:  # make a dummy METADATA file
            f.write("Name: " + package + "\n")
            f.write("Version: " + version + "\n")

        with open(str(path_distinfo / "INSTALLER"), "w") as f:  # make a dummy METADATA file
            f.write("github\n")
        with open(str(path_distinfo / "RECORD"),"w") as f:
            pass  # just to create the file to be recorded

        with open(str(path_distinfo / "RECORD"), "w") as record_file:

            for p in (path, path_distinfo):
                for file in p.glob("**/*"):

                    if file.is_file():
                        name = file.relative_to(sitepackages_path).as_posix()  # make sure we have slashes
                        record_file.write(name + ",")

                        if (file.stem == "RECORD" and p == path_distinfo) or ("__pycache__" in name.lower()):
                            record_file.write(",")
                        else:
                            with file.open("rb") as f:
                                file_contents = f.read()
                                hash = "sha256=" + base64.urlsafe_b64encode(
                                    hashlib.sha256(file_contents).digest()
                                ).decode("latin1").rstrip("=")
                                # hash calculation derived from wheel.py in pip

                                length = str(len(file_contents))
                                record_file.write(hash + "," + length)

                        record_file.write("\n")

    return info

if __name__ == "__main__":
    info = _install(
        files="pil_printer.py !changelog.txt".split()
    )
    print(info.package + " " + info.version + " successfully installed in " + info.path)
    print("files copied: ", ", ".join(info.files_copied))


pil_printer.py
import win32print
import win32ui
from PIL import Image
from PIL import ImageWin
from pathlib import Path
import collections

"""
pil_printer

To be used to print a (or many) images to a printer.
Only Windows is supported.

Based on http://timgolden.me.uk/python/win32_how_do_i/print.html
"""
__version__ = "1.0.0"


def spec(image, width=None, height=None, horizontal_offset=None, vertical_offset=None, unit=None):
    """
    Prepares an image to be printed with custom properties
    
    Parameters
    ----------
    image : PIL image
        Image to be used
        
    width : float 
        Width of the printed image in the given unit
        Defaults to width given in call to pil_printer
        
    height : float 
        Height of  the printed image in the given unit
        Defaults to height given in call to pil_printer        
                    
    horizontal_offset : float
        Defaults to horizontal_offset given in call to pil_printer
        
    vertical_offset : float
        Defaults to vertical_offset given in call to pil_printer
        
    unit : str
        Defaults to unit given in call to pil_printer
        
    Returns
    -------
    spec : dict
        to be used as image in call to pil_printer
    """
    return dict(image=image, width=width, height=height, horizontal_offset=horizontal_offset, vertical_offset=vertical_offset, unit=unit)


def pil_printer(*image, width=None, height=None, horizontal_offset=0, vertical_offset=0, unit="cm", printer_name=None):
    """
    Prints a PIL image to a printer
    
    Parameters
    ----------
    image : PIL image
        Image to be printed
        Or it is possible to specify many images. However this nearly always requires 
        the images to be given as a spec, in order to set offset(s) and possibly height/width
        
    width : float 
        Width of the printed image in the given unit
        No default
        
    height : float 
        Height of  the printed image in the given unit
        No default
                    
    horizontal_offset : float
        Normally the image is printed centered on the page.
        With this parameter, the image can be offset horizontally
        Default: 0
        
    vertical_offset : float
        Normally the image is printed centered on the page.
        With this parameter, the image can be offset vertically
        Default: 0        
        
    unit : str
        Either 'cm' (default) or 'inch'

    printer : str
        Printer name to used
        Default: default printer

    Note
    ----
    One and only one of width/height has to be specified
    Only compatible with Windows platforms.
    """
    ImageInfo = collections.namedtuple("ImageInfo", "image x0 y0 x1 y1")

    LOGPIXELSX = 88
    LOGPIXELSY = 90

    PHYSICALWIDTH = 110
    PHYSICALHEIGHT = 111

    if printer_name is None:
        printer_name = win32print.GetDefaultPrinter()

    image_infos = []
    hDC = win32ui.CreateDC()
    hDC.CreatePrinterDC(printer_name)
    printer_size = hDC.GetDeviceCaps(PHYSICALWIDTH), hDC.GetDeviceCaps(PHYSICALHEIGHT)
    pixels_per_inch = hDC.GetDeviceCaps(LOGPIXELSX), hDC.GetDeviceCaps(LOGPIXELSY)

    width_default = width
    height_default = height
    horizontal_offset_default = horizontal_offset
    vertical_offset_default = vertical_offset
    unit_default = unit

    for image in image[:]:
        if isinstance(image, dict):
            width = width_default if image["width"] is None else image["width"]
            height = height_default if image["height"] is None else image["height"]
            horizontal_offset = horizontal_offset_default if image["horizontal_offset"] is None else image["horizontal_offset"]
            vertical_offset = vertical_offset_default if image["vertical_offset"] is None else image["vertical_offset"]
            unit = unit_default if image["unit"] is None else image["unit"]
            image = image["image"]
        else:
            width = width_default
            height = height_default
            horizontal_offset = horizontal_offset_default
            vertical_offset = vertical_offset_default
            unit = unit_default

        size = [None, None]
        if width is not None:
            size[0] = width

        if height is not None:
            size[1] = height

        if all(i_size is None for i_size in size):
            raise ValueError("neither width nor height specified")

        if not any(i_size is None for i_size in size):
            raise ValueError("width and he`ight specified")

        offset = [horizontal_offset, vertical_offset]

        if unit == "cm":
            inches_in_unit = 2.54
        elif unit == "inch":
            inches_in_unit = 1
        else:
            raise ValueError(f"unit should be 'cm' or 'inch', not {repr(unit)}")

        offset_in_pixels = [offset[i] * pixels_per_inch[i] / inches_in_unit for i in (0, 1)]

        if isinstance(image, (str, Path)):
            image = Image.open(image)

        if not isinstance(image, Image.Image):
            raise ValueError("image is not a proper Image.Image")

        for i in (0, 1):
            if size[i] is not None:
                scale = size[i] * pixels_per_inch[i] / image.size[i] / inches_in_unit

        scaled_width, scaled_height = [int(scale * i_size) for i_size in image.size]
        x0 = int(offset_in_pixels[0] + (printer_size[0] - scaled_width) / 2)
        y0 = int(offset_in_pixels[1] + (printer_size[1] - scaled_height) / 2)
        x1 = x0 + scaled_width
        y1 = y0 + scaled_height
        image_infos.append(ImageInfo(image, x0, y0, x1, y1))

    hDC.StartDoc("pil_printer")
    hDC.StartPage()

    for image_info in image_infos:
        dib = ImageWin.Dib(image_info.image)
        dib.draw(hDC.GetHandleOutput(), (image_info.x0, image_info.y0, image_info.x1, image_info.y1))

    hDC.EndPage()
    hDC.EndDoc()
    hDC.DeleteDC()


if __name__ == "__main__":
    from PIL import ImageDraw

    im = Image.new(mode="RGBA", size=(1000, 1000), color=(255, 255, 255))
    d = ImageDraw.Draw(im=im)
    for i in range(1, 10):
        d.rectangle(xy=(i * 50, i * 50, 1000 - i * 50, 1000 - i * 50), outline=(i * 25, i * 25, 0))

    pil_printer(spec(im, horizontal_offset=2, width=5), spec(im, vertical_offset=10), width=5, horizontal_offset=0, vertical_offset=4, unit="cm")
